tools/clang/tools/dxcvalidator/dxcvalidator.cpp

```diff
@@ -47,12 +48,50 @@ struct DiagRestore {
   ~DiagRestore() { Ctx.setDiagnosticHandler(OrigHandler, OrigDiagContext); }
 };
 
+static void HashAndUpdate(DxilContainerHeader *Container) {
+  // Compute hash and update stored hash.
+  // Hash the container from this offset to the end.
+  static const uint32_t DXBCHashStartOffset =
+      offsetof(struct DxilContainerHeader, Version);
+  const unsigned char *DataToHash =
+      (const unsigned char *)Container + DXBCHashStartOffset;
+  unsigned AmountToHash = Container->ContainerSizeInBytes - DXBCHashStartOffset;
+  ComputeHashRetail(DataToHash, AmountToHash, Container->Hash.Digest);
+}
+
+static void HashAndUpdateOrCopy(uint32_t Flags, IDxcBlob *Shader,
+                                IDxcBlob **Hashed) {
+  if (Flags & DxcValidatorFlags_InPlaceEdit) {
+    HashAndUpdate((DxilContainerHeader *)Shader->GetBufferPointer());
+    *Hashed = Shader;
+    Shader->AddRef();
+  } else {
+    // Possible gotcha: the blob allocated here is tied to this .dll, so the
+    // DLL shouldn't be unloaded before the blob is released.
+    CComPtr<AbstractMemoryStream> HashedBlobStream;
+    uint32_t HR =
+        CreateMemoryStream(DxcGetThreadMallocNoRef(), &HashedBlobStream);
+    if (FAILED(HR))
+      throw hlsl::Exception(HR);
+
+    unsigned long CB;
+    HR = HashedBlobStream->Write(Shader->GetBufferPointer(),
+                                 Shader->GetBufferSize(), &CB);
+    if (FAILED(HR))
+      throw hlsl::Exception(HR);
+    HashAndUpdate((DxilContainerHeader *)HashedBlobStream->GetPtr());
+    HR = HashedBlobStream.QueryInterface(Hashed);
+    if (FAILED(HR))
+      throw hlsl::Exception(HR);
+  }
+}
+
 static uint32_t runValidation(
     IDxcBlob *Shader,
     uint32_t Flags,            // Validation flags.
     llvm::Module *Module,      // Module to validate, if available.
     llvm::Module *DebugModule, // Debug module to validate, if available
-    AbstractMemoryStream *DiagMemStream) {
+    AbstractMemoryStream *DiagMemStream, IDxcBlob **Hashed) {
```

Comment:
This `Hashed` output pointer is not written to in this function, nor does it compute the hash.

This function is for the internal case, which I think needs an overhaul, especially now that we are integrating the hash here.

It's likely best if we did the overhaul as a separate change before integrating the hash changes.  It should greatly simplify the scenarios.

We should get rid of the special internal validator case where it accepts the main module as a `llvm::Module` and always validate the serialized bitcode in the container.  We will want to keep the debug module argument, so it may be passed along internally, avoiding extra deserialization, since it's only used for diagnostic locations, and not the thing being validated.

Reply:
Ok, so this is really meant to be the inner function used by both code paths that have debug bitcode and the internal path that has a debug module.  We should probably use this in the other `runValidation` overload, and in the case of `DxcValidatorFlags_ModuleOnly` use a different function entirely to call this one which deserialized the debug module and does not have a `Hashed` output at all.

Reply:
Removed.